cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/BasicServiceWorkerContext.java

```diff
@@ -95,10 +96,17 @@ public BasicServiceWorkerContext(ServiceWorkerSpecification spec, Program progra
     this.userMetrics = new ProgramUserMetrics(getMetricCollector(metricsSerice, program,
                                                                  spec.getName(), runId.getId(), instanceId));
     this.runtimeArgs = runtimeArgs.asMap();
+
+    // The cache expiry should be greater than (2 * transaction.timeout) and at least 2 minutes.
+    // This ensures that when a dataset instance is requested multiple times during a single transaction,
+    // the same instance is always returned.
```

Comment:
The context lives across multiple transactions. However, I'm not sure what you mean. If it expires during the second transaction a new instance is given - which is not a problem. A problem arises only when 2 different instance of the same dataset are given during the same transaction.

Unless I misunderstand what you're saying?